include/pybind11/stl_bind.h

```diff
@@ -364,6 +364,36 @@ NAMESPACE_BEGIN(detail)
 /* Fallback functions */
 template <typename, typename, typename... Args> void map_if_insertion_operator(const Args&...) { }
 
+template <typename Map, typename Class_, typename... Args> void map_if_const_value(Class_ &cl, const Args&...)
+{
+	using KeyType = typename Map::key_type;
+	using MappedType = typename Map::mapped_type;
+
+	cl.def("__setitem__",
+		   [](Map &m, const KeyType &k, const MappedType &v) {
+			   auto it = m.find(k);
+			   if (it != m.end()) it->second = v;
+			   else m.emplace(k, v);
+		   });
+
+}
+
+template<typename Map, typename Class_, typename std::enable_if<std::is_const<typename Map::mapped_type>::value, int>::type = 0>
```

Comment:
I think you could generalize this a bit by changing
`std::is_const<typename Map::mapped_type>::value`
to
`!std::is_copy_assignable<Map::mapped_type>::value`
which would still catch `const` values, but would also catch non-const types that don't have a copy assignment operator.